### PR TITLE
[Jballerina] Add codegen for futures

### PR DESCRIPTION
**Pull request by vinok88**
```
future<int> p = start foo("abc", 7);
```

> Then this lambda function is handed over to the Scheduler to be scheduled. Current scheduler implementation is not finalized, added one to get this working only. 

> Scheduler returns the strand instance that the async function is running.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        if (executorService == null) {
+            executorService = Executors.newFixedThreadPool(3);
+        }
+        AsyncExecutor async = new AsyncExecutor(function, params);
+        strand.future = executorService.submit(async);
+
+        return strand;
+    }
+
+    public static void shutDown() {
```
Comment (Kishanthan) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java]: name should be "shutdown"  ?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ * @since 0.995.0
+ */
+public class Scheduler {
+
+    public static ExecutorService executorService = null;
+
+    public static Strand schedule(Object[] params, Function function) {
+        Strand strand = new Strand();
+        // find the return type based on type
+        if (executorService == null) {
```
Comment (Kishanthan) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java]: Why do we have to check null here?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            executorService.shutdown();
+        }
+    }
+
+}
+
+/**
+ * A callable implementation to invoke a given ballerina function asynchronously.
+ * @param <T>
+ */
```
Comment (Kishanthan) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java]: Missing version

File: stdlib/jvm/src/main/java/org/ballerinalang/nativeimpl/jvm/methodvisitor/VisitInvokeDynamicInsn.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        MethodVisitor mv = ASMUtil.getRefArgumentNativeData(context, 0);
+        mv.visitInvokeDynamicInsn("apply", "()Ljava/util/function/Function;",
+                new Handle(Opcodes.H_INVOKESTATIC, "java/lang/invoke/LambdaMetafactory",
+                        "metafactory", "(Ljava/lang/invoke/MethodHandles$Lookup;"
+                        + "Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodType;"
+                        + "Ljava/lang/invoke/MethodHandle;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/CallSite;",
+                        false), new Object[]{Type.getType("(Ljava/lang/Object;)Ljava/lang/Object;"),
+                        new Handle(Opcodes.H_INVOKESTATIC, className, lambdaName,
+                                "([Ljava/lang/Object;)Ljava/lang/Object;", false),
+                        Type.getType("([Ljava/lang/Object;)Ljava/lang/Object;")});
```
Comment (Kishanthan) [inline stdlib/jvm/src/main/java/org/ballerinalang/nativeimpl/jvm/methodvisitor/VisitInvokeDynamicInsn.java]: Do we have already available constants for these strings?

File: tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal [hunk 5]
```diff
@@ -213,3 +213,27 @@ function acceptAnydata(anydata data) returns anydata {
...
+
+public function foo2(string k) returns string {
+	global = global + 100;
+    return k;
+}
+
+public function acceptFuture(future<string> pp) returns future<string> {
+	future<string> p2 = pp;
+	return p2;
+}
```
Comment (Kishanthan) [inline tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal]: Above new functions additions are not formatted properly

File: stdlib/jvm/src/main/java/org/ballerinalang/nativeimpl/jvm/methodvisitor/VisitInvokeDynamicInsn.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import org.objectweb.asm.Type;
+
+import static org.ballerinalang.model.types.TypeKind.OBJECT;
+import static org.ballerinalang.model.types.TypeKind.STRING;
+import static org.ballerinalang.nativeimpl.jvm.ASMUtil.JVM_PKG_PATH;
+import static org.ballerinalang.nativeimpl.jvm.ASMUtil.METHOD_VISITOR;
+
+/**
+ * Code generation for invoke dynamic instruction.
+ */
```
Comment (Kishanthan) [inline stdlib/jvm/src/main/java/org/ballerinalang/nativeimpl/jvm/methodvisitor/VisitInvokeDynamicInsn.java]: Missing version

File: tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal [hunk 7]
```diff
@@ -213,3 +213,27 @@ function acceptAnydata(anydata data) returns anydata {
    ia[1] = 1000;
    return ia[1];
 }
+
+int global = 0;
```
Comment (warunalakshitha) [inline tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal]: I think its better to move future tests to separate bal.

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_terminator_gen.bal [hunk 8]
```diff
@@ -18,11 +18,15 @@ type TerminatorGenerator object {
     jvm:MethodVisitor mv;
     BalToJVMIndexMap indexMap;
     LabelGenerator labelGen;
+    int lambdaIndex = 0;
+    bir:Package module;
 
-    public function __init(jvm:MethodVisitor mv, BalToJVMIndexMap indexMap, LabelGenerator labelGen) {
+
+    public function __init(jvm:MethodVisitor mv, BalToJVMIndexMap indexMap, LabelGenerator labelGen, bir:Package module) {
```
Comment (warunalakshitha) [inline compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_terminator_gen.bal]: Length of this line seems to be  more than 120 characters..

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java (hunk 2 above)
Comment (vinok88) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java]: Will move the executor init part out of schedule method. Then we do not need the null check.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java (hunk 1 above)
Comment (vinok88) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java]: Fixed. Thanks.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java (hunk 2 above)
Comment (vinok88) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java]: Fixed. Thanks.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java (hunk 3 above)
Comment (vinok88) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/Scheduler.java]: Fixed. Thanks.

File: stdlib/jvm/src/main/java/org/ballerinalang/nativeimpl/jvm/methodvisitor/VisitInvokeDynamicInsn.java (hunk 4 above)
Comment (vinok88) [reply stdlib/jvm/src/main/java/org/ballerinalang/nativeimpl/jvm/methodvisitor/VisitInvokeDynamicInsn.java]: No. Will add for common strings.

File: tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal (hunk 7 above)
Comment (vinok88) [reply tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal]: Why? Future also a type.

File: tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal (hunk 5 above)
Comment (vinok88) [reply tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal]: Fixed thanks.

File: stdlib/jvm/src/main/java/org/ballerinalang/nativeimpl/jvm/methodvisitor/VisitInvokeDynamicInsn.java (hunk 6 above)
Comment (vinok88) [reply stdlib/jvm/src/main/java/org/ballerinalang/nativeimpl/jvm/methodvisitor/VisitInvokeDynamicInsn.java]: Fixed thanks.

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal [hunk 9]
```diff
@@ -352,6 +364,49 @@ function generateMethod(bir:Function func, jvm:ClassWriter cw, bir:Package modul
     mv.visitEnd();
 }
 
+function generateLambdaMethod(bir:Call callIns, jvm:ClassWriter cw, string className, string lambdaName) {
+    jvm:MethodVisitor mv = cw.visitMethod(ACC_PUBLIC + ACC_STATIC, lambdaName, 
+                                "([Ljava/lang/Object;)Ljava/lang/Object;", (), ());
```
[the remark this answers is not in the document]
Comment (vinok88) [reply compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal]: Fixed, Thanks.

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_terminator_gen.bal (hunk 8 above)
Comment (vinok88) [reply compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_terminator_gen.bal]: Fixed, Thanks.

File: tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal (hunk 7 above)
Comment (warunalakshitha) [reply tests/ballerina-unit-test/src/test/resources/test-src/jvm/types.bal]: May be once we have other operations with futures better to move to new file. +1 for now to use same types.bal.

**Comment by Kishanthan** [comment]
conflicts :( shall we fix them? @vinok88

**Pull request merged**